src/kubernetes/api_watcher.rs

```diff
@@ -152,3 +164,457 @@ pub mod invocation {
+                                "reason": "Puter go BOOM",
+                                "status": "Failure"
+                            }
+                        }"#,
+                        );
+                }),
+                vec![Box::new(|item| {
+                    assert_watch_response(
+                        item,
+                        WatchEvent::ErrorStatus(Status {
+                            code: Some(500),
+                            message: Some("Internal Server Error".to_owned()),
+                            reason: Some("Puter go BOOM".to_owned()),
+                            status: Some("Failure".to_owned()),
+                            ..Default::default()
+                        }),
+                    );
+                })],
+            ),
+            // Bad JSON from API
+            (
+                Box::new(|when, then| {
+                    when.method(GET).path("/api/v1/pods");
+                    then.status(200)
+                        .header("Content-Type", "application/json")
+                        .body(
+                            r#"
+                            "type": "ADDED",
+                            "object":
+                                "kind": "Pod",
+                                "apiVersion": "v1",
+                                "metadata": {
+                                    "uid": "uid0"
+                                }
+                            }
+                         }"#,
+                        );
+                }),
+                vec![Box::new(|item| {
+                    let error = item.unwrap_err();
+                    match error {
+                        watcher::stream::Error::Other { .. } => {}
+                        _ => panic!("Expect an 'other' error"),
+                    }
+                })],
+            ),
+            // Incorrect object type
+            (
+                Box::new(|when, then| {
+                    when.method(GET).path("/api/v1/pods");
+                    then.status(200)
+                        .header("Content-Type", "application/json")
+                        .body(
+                            r#"
+                            "type": "MODIFIED",
+                            "object": {
+                                "kind": "StatefulSet",
+                                "apiVersion": "v1",
+                                "metadata": {
+                                    "uid": "uid0"
+                                }
+                            }
+                         }"#,
+                        );
+                }),
+                vec![Box::new(|item| {
+                    let error = item.unwrap_err();
+                    match error {
+                        watcher::stream::Error::Other { .. } => {}
+                        _ => panic!("Expect an 'other' error"),
+                    }
+                })],
+            ),
```

Comment:
I'd expect a watch response, rather than a stream-level error here, interesting that tests are passing. Something might be off. https://github.com/timberio/vector/pull/6053/files#r563191159 should help with resolving this.